File: framework/fit/python/fit_flowable/subscriber/abstract_subscriber.py
# -- encoding: utf-8 --
# Copyright (c) Huawei Technologies Co., Ltd. 2024-2024. All rights reserved.
"""
功 能：订阅者的抽象父类。
"""
from typing import Optional

from fit_flowable.core.exception import FlowableException
from fit_flowable.core.subscriber import Subscriber
from fit_flowable.core.subscription import Subscription
from fit_flowable.util.concurrent import AtomicBool


class AbstractSubscriber(Subscriber):
    """
    表示订阅者的抽象父类。
    """

    def __init__(self):
        self._pre_subscription: Optional[Subscription] = None
        self._completed: AtomicBool = AtomicBool()
        self._failed: AtomicBool = AtomicBool()

    def do_on_subscribed(self) -> None:
        """
        订阅关系发生时的具体行为。
        """
        pass

    def do_consume(self, data) -> None:
        """
        消费数据时的具体行为。

        :param data: 待消费的数据。
        """
        pass

    def do_fail(self, cause: Exception) -> None:
        """
        消费异常终结信号时的具体行为。

        :param cause: 待消费的异常。
        """
        pass

    def do_complete(self) -> None:
        """
        消费正常终结信号时的具体行为。
        """
        pass

    def get_pre_subscription(self) -> Subscription:
        """
        获取上游的订阅关系。

        :return: 上游的订阅关系。
        """
        if self._pre_subscription is None:
            raise FlowableException("the previous subscription is not available before subscribed.")
        return self._pre_subscription

    def on_subscribed(self, subscription: Subscription) -> None:
        if self._pre_subscription is not None:
            raise FlowableException("the subscriber cannot be subscribed twice.")
        if subscription is None:
            raise ValueError("the subscription cannot be none.")
        self._pre_subscription: Subscription = subscription
        self.do_on_subscribed()

    def consume(self, data) -> None:
        if self._completed.get() or self._failed.get():
            return
        self.do_consume(data)

    def fail(self, cause: Exception) -> None:
        if self._failed.get() or not self._completed.compare_and_set(False, True):
            return
        self.do_fail(cause)

    def complete(self) -> None:
        if self._completed.get() or not self._failed.compare_and_set(False, True):
            return
        self.do_complete()

File: framework/fit/java/fit-reactor/src/main/java/modelengine/fitframework/flowable/subscriber/AbstractSubscriber.java
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) 2024 Huawei Technologies Co., Ltd. All rights reserved.
 *  This file is a part of the ModelEngine Project.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

package modelengine.fitframework.flowable.subscriber;

import static modelengine.fitframework.inspection.Validation.isTrue;
import static modelengine.fitframework.inspection.Validation.notNull;

import modelengine.fitframework.flowable.FlowableException;
import modelengine.fitframework.flowable.Subscriber;
import modelengine.fitframework.flowable.Subscription;
import modelengine.fitframework.inspection.Nonnull;

import java.util.concurrent.atomic.AtomicBoolean;

/**
 * 表示 {@link Subscriber} 的抽象父类。
 *
 * @param <T> 表示订阅的数据类型的 {@link T}。
 * @author 季聿阶
 * @since 2024-02-10
 */
public abstract class AbstractSubscriber<T> implements Subscriber<T> {
    private Subscription preSubscription;
    private final AtomicBoolean completed = new AtomicBoolean();
    private final AtomicBoolean failed = new AtomicBoolean();

    @Override
    public void onSubscribed(Subscription subscription) {
        isTrue(this.preSubscription == null, () -> new FlowableException("The subscriber cannot be subscribed twice."));
        this.preSubscription = notNull(subscription, "The subscription cannot be null.");
        this.onSubscribed0(this.preSubscription);
    }

    /**
     * 表示订阅关系发生时的事件。
     *
     * @param subscription 表示给定的订阅关系的 {@link Subscription}。
     */
    protected abstract void onSubscribed0(@Nonnull Subscription subscription);

    /**
     * 获取上游的订阅关系。
     *
     * @return 表示上游的订阅关系的 {@link Subscription}。
     */
    @Nonnull
    protected Subscription getPreSubscription() {
        return notNull(this.preSubscription,
                () -> new FlowableException("The previous subscription is not available before subscribed."));
    }

    @Override
    public void consume(T data) {
        if (this.completed.get() || this.failed.get()) {
            return;
        }
        this.consume(this.preSubscription, data);
    }

    /**
     * 消费订阅的数据。
     *
     * @param subscription 表示对上游的订阅关系的 {@link Subscription}。
     * @param data 表示待消费的数据的 {@link T}。
     */
    protected abstract void consume(@Nonnull Subscription subscription, T data);

    @Override
    public void complete() {
        if (this.failed.get() || !this.completed.compareAndSet(false, true)) {
            return;
        }
        this.complete(this.preSubscription);
    }

    /**
     * 表示正常终结信号。
     *
     * @param subscription 表示对上游的订阅关系的 {@link Subscription}。
     */
    protected abstract void complete(@Nonnull Subscription subscription);

    @Override
    public boolean isCompleted() {
        return this.completed.get();
    }

    @Override
    public void fail(Exception cause) {
        if (this.completed.get() || !this.failed.compareAndSet(false, true)) {
            return;
        }
        this.fail(this.preSubscription, cause);
    }

    /**
     * 表示异常终结信号。
     *
     * @param subscription 表示对上游的订阅关系的 {@link Subscription}。
     * @param cause 表示订阅中的错误信息的 {@link Exception}。
     */
    protected abstract void fail(@Nonnull Subscription subscription, Exception cause);

    @Override
    public boolean isFailed() {
        return this.failed.get();
    }
}
